server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java

```diff
@@ -151,6 +152,120 @@ void throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize() {
+
+            Flux<Integer> originalFlux = Flux.range(0, 10);
+            Function<Integer, Publisher<Integer>> longRunningOperation =
+                i -> {
+                    if (i == 5) {
+                        return Mono.error(new RuntimeException());
+                    }
+                    return Mono.just(i);
+                };
+
+            List<Integer> results = originalFlux
+                .transform(ReactorUtils.<Integer, Integer>throttle()
+                    .elements(windowMaxSize)
+                    .per(windowDuration)
+                    .forOperation(longRunningOperation))
+                .collectList()
+                .block();
+
+            assertThat(results)
+                .containsExactly(0, 1, 2, 3, 4, 6, 7, 8, 9);
+        }
+
+        @Test
+        void throttleShouldNotAbortProcessingUponUpstreamError() {
+            Hooks.onOperatorDebug();
+
+            int windowMaxSize = 3;
+            Duration windowDuration = Duration.ofMillis(100);
+
+            Flux<Integer> originalFlux = Flux.range(0, 10)
+                .flatMap(i -> {
+                    if (i == 5) {
+                        return Mono.error(new RuntimeException());
+                    }
+                    return Mono.just(i);
+                });
+
+            List<Integer> results = originalFlux
+                .transform(ReactorUtils.<Integer, Integer>throttle()
+                    .elements(windowMaxSize)
+                    .per(windowDuration)
+                    .forOperation(Mono::just))
+                .collectList()
+                .block();
+
+            assertThat(results)
+                .containsExactly(0, 1, 2, 3, 4, 6, 7, 8, 9);
+        }
+
+        @Test
+        void throttleShouldHandleLargeFluxes() {
+            int windowMaxSize = 2;
+            Duration windowDuration = Duration.ofMillis(1);
+
+            Flux<Integer> originalFlux = Flux.range(0, 10000);
+
+            originalFlux
+                .transform(ReactorUtils.<Integer, Integer>throttle()
+                    .elements(windowMaxSize)
+                    .per(windowDuration)
+                    .forOperation(Mono::just))
+                .blockLast();
+        }
+
+        @Test
+        void throttleShouldGenerateSmallerWindowsWhenUpstreamIsSlow() {
+            int windowMaxSize = 3;
+            Duration windowDuration = Duration.ofMillis(20);
+            Stopwatch stopwatch = Stopwatch.createUnstarted();
+
+            Flux<Long> originalFlux = Flux.interval(Duration.ofMillis(10));
+
+            ImmutableList<Long> perWindowCount = originalFlux
+                .transform(ReactorUtils.<Long, Long>throttle()
+                    .elements(windowMaxSize)
+                    .per(windowDuration)
+                    .forOperation(i -> Mono.fromCallable(() -> stopwatch.elapsed(TimeUnit.MILLISECONDS))))
+                .map(i -> i / 20)
+                .doOnSubscribe(signal -> stopwatch.start())
```

Comment:
what is the usage of stopwatch if we don't assert on it?

Reply:
stopWatch is here to tell "which window" an element belongs to.